src/scoping/langs/c.rs

```diff
@@ -0,0 +1,130 @@
+    /// Function declaration.
+    FunctionDecl,
+    /// `switch` blocks.
+    Switch,
+    /// `if` blocks.
+    If,
+    /// `for` blocks.
+    For,
+    /// `while` blocks.
+    While,
+    /// `do` blocks.
+    Do,
+    /// `union` blocks.
+    Union,
+    /// Identifier.
+    Identifier,
+    /// Declaration.
+    Declaration,
+    /// Call expression.
+    CallExpression,
+}
+
+impl From<PreparedCQuery> for TSQuery {
+    fn from(value: PreparedCQuery) -> Self {
+        Self::new(
+            &C::lang(),
+            match value {
+                PreparedCQuery::Comments => "(comment) @comment",
+                PreparedCQuery::Strings => "[(string_literal) (system_lib_string)] @string",
+                PreparedCQuery::Includes => "(preproc_include) @include",
+                PreparedCQuery::TypeDef => "(type_definition) @typedef",
+                PreparedCQuery::Enum => "(enum_specifier) @enum",
+                PreparedCQuery::Struct => "(struct_specifier) @struct",
+                PreparedCQuery::Variable => "(declaration) @var",
+                PreparedCQuery::FunctionDef => "(function_definition) @function_definition",
+                PreparedCQuery::FunctionDecl => "(function_declarator) @function_decl",
+                PreparedCQuery::Switch => "(switch_statement) @switch",
+                PreparedCQuery::If => "(if_statement) @if",
+                PreparedCQuery::For => "(for_statement) @for",
+                PreparedCQuery::While => "(while_statement) @while",
+                PreparedCQuery::Union => "(union_specifier) @union",
+                PreparedCQuery::Do => "(do_statement) @do",
+                PreparedCQuery::Identifier => "(identifier) @ident",
+                PreparedCQuery::Declaration => "(declaration) @decl",
+                PreparedCQuery::CallExpression => "(call_expression) @call",
+            },
+        )
+        .expect("Prepared queries to be valid")
+    }
+}
+
+/// A custom tree-sitter query for C.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CustomCQuery(String);
+
+impl FromStr for CustomCQuery {
+    type Err = QueryError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match TSQuery::new(&C::lang(), s) {
+            Ok(_) => Ok(Self(s.to_string())),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl From<CustomCQuery> for TSQuery {
+    fn from(value: CustomCQuery) -> Self {
+        Self::new(&C::lang(), &value.0)
+            .expect("Valid query, as object cannot be constructed otherwise")
+    }
+}
+
+impl LanguageScoper for C {
+    fn lang() -> TSLanguage {
+        tree_sitter_c::LANGUAGE.into()
+    }
+
+    fn pos_query(&self) -> &TSQuery {
+        &self.positive_query
+    }
+
+    fn neg_query(&self) -> Option<&TSQuery> {
+        self.negative_query.as_ref()
+    }
+}
+
+impl Find for C {
+    fn extensions(&self) -> &'static [&'static str] {
+        &["c", "h"]
+    }
+
+    fn is_path_invalid(&self, _: &Path) -> bool {
```

Comment:
You can remove `is_path_invalid` altogether: it has a default implementation which does the exact same:

https://github.com/alexpovel/srgn/blob/45b64556df1abb983bd067112dacdbb645dd9b47/src/find/mod.rs#L31